### ISSUE TITLE
Please add some backoff or pause for recursive uploads

### ISSUE THREAD

**Issue by felliott** [osfclient/osfclient#103]
Hey folks!

Our (COS's) WaterButler servers are getting hammered by recursive uploads from osf-cli and are having a hard time keeping up.  Could you add in a delay between requests or a backoff when you encounter an error?  Thank you!

Cheers,
@felliott

**Comment by betatim** [comment]
Sure thing. Is there a way to tell that too many requests are being made from the response water butler/osf.io sends? Or should we implement a generic upper limit of no more than X requests per second?

---

The right place to implement a rate limit would be in [`OSFSession`](https://github.com/dib-lab/osf-cli/blob/007a1b7a50576e568e1ef446f5e4f20ce30e0f79/osfclient/models/session.py#L6) through which all requests go.

**Comment by felliott** [comment]
'Fraid not.  It's something we need to add, but we don't have it yet.  A generic limit would be good.  I don't have a sense of what would be good, shall we start at 1 / sec?

**Comment by felliott** [comment]
Hey @betatim,

Apologies!  It appears I've slandered the good name of `osf-cli`.  I've investigated deeper, and the load we were experiencing yesterday was from a user running their own bespoke script.  Some poor `osf-cli` user was caught in the flurry of errors, which set me on the wrong track.  That said, I think this is still a useful feature, especially as this project grows in popularity.  #105 looks great to me.  Thank you!

Cheers,
@felliott

**Comment by betatim** [comment]
Good to know.

I did a bit of "measuring" how often you end up being limited by the 1sec rule and it happens "semi often". Looks like you get to sleep for 0.4sec or so. Are there some requests which are particularly costly on the server side?

Maybe in a future iteration we can rate limit those but keep requests that (for example) list files unthrottled to make it feel as fast as possible.

**Comment by felliott** [comment]
Only limiting uploads sounds reasonable to me.  Listing should generally be pretty fast.  Another thing that would help would be if `osf-cli` would check the compute hashes before uploading files and skip uploading if they match.  `osfstorage` includes md5, sha1, and sha256 hashes in the file metadata.  Most providers also provide some sort of hash, but which ones varies between providers.

**Comment by betatim** [comment]
Computing and comparing hashes is also relevant to #79 and other "sync" related commands.

**Comment by mfraezz** [comment]
> Apologies! It appears I've slandered the good name of `osf-cli`. I've investigated deeper, and the load we were experiencing yesterday was from a user running their own bespoke script.

This time, it's real. We have a user syncing approximately 1200 files (specifically downloading, but I imagine they might also upload later) at a rate of approximately 5 requests per second, and this is causing performance degradation (slowness, not total failure) in the API containers. This has happened multiple times tonight, I suspect that one of the requests may have timed out and killed his `osfclient`.

The `User-Agent` for these requests is `osfclient v0.0.1`, as found [here](https://github.com/osfclient/osfclient/blob/44b9a87e8c1ae6b63cdecd27a924af3fc2bf94cf/osfclient/models/session.py#L21), so I believe this is not a false allegation. _Aside: should that be formatting in `__version__`? I noticed the current version is `0.0.3`._

I've made a note to analyze the queries tomorrow to determine how those endpoints should be optimized, but would it be possible to merge #105 in, maybe with a version bump? We can contact the user and request that they upgrade.

We currently have a (rather lax) throttling mechanism for the API, and could add a rule for either `osfclient` or "burst"-rates generally, but this would generate `429 Too Many Requests`, leading to more problems if they're not handled appropriately here (which would involve parsing the `json` response and sleeping for `X` seconds, see https://github.com/encode/django-rest-framework/blob/3.8.2/rest_framework/exceptions.py#L220-L238 ).
